src/sage/combinat/k_regular_sequence.py

```diff
@@ -918,6 +1026,265 @@ def _n_to_index_(self, n):
+
+        @cached_function(key=lambda lines: len(lines))  # we assume that existing lines are not changed (we allow appending of new lines)
+        def some_inverse_U_matrix(lines):
+            d = len(seq(0)) + len(lines)
+
+            for m_indices in cantor_product(xsrange(n_max), repeat=d, min_slope=1):
+                U = Matrix(domain, d, d, [values(m, lines) for m in m_indices]).transpose()
+                try:
+                    return U.inverse(), m_indices
+                except ZeroDivisionError:
+                    pass
+            else:
+                raise RuntimeError
+
+        def guess_linear_dependence(t_L, r_L, lines):
+            iU, m_indices = some_inverse_U_matrix(lines)
+            X_L = vector(f(k**t_L * m + r_L) for m in m_indices)
+            return X_L * iU
+
+        def verify_linear_dependence(t_L, r_L, linear_dependence, lines):
+            return all(f(k**t_L * m + r_L) ==
+                       linear_dependence * vector(values(m, lines))
+                       for m in xsrange(0, (n_max - r_L) // k**t_L + 1))
+
+        def find_linear_dependence(t_L, r_L, lines):
+            linear_dependence = guess_linear_dependence(t_L, r_L, lines)
+            if not verify_linear_dependence(t_L, r_L, linear_dependence, lines):
+                raise ValueError
+            return linear_dependence
+
+        left = None
+        if seq(0):
+            try:
+                solution = find_linear_dependence(0, 0, [])
+            except ValueError:
+                pass
+            else:
+                left = vector(solution)
+
+        to_branch = []
+        lines = []
+        def include(line):
+            to_branch.append(line)
+            lines.append(line)
+            t, r, s = line
+            logger.info('including f_{%s*m+%s}', k**t, r)
+
+        if left is None:
+            line_L = (0, 0, 0)  # entries (t, r, s) --> k**t * m + r, belong to M_s
+            include(line_L)
+            left = vector((len(seq(0)) + len(lines)-1)*(zero,) + (one,))
+
+        while to_branch:
+            line_R = to_branch.pop(0)
+            t_R, r_R, s_R = line_R
+            if t_R >= max_dimension:
+                raise RuntimeError
+
+            t_L = t_R + 1
+            for s_L in srange(k):
+                r_L = k**t_R * s_L + r_R
+                line_L = t_L, r_L, s_L
+
+                try:
+                    solution = find_linear_dependence(t_L, r_L, lines)
+                except ValueError:
+                    include(line_L)
+                    solution = (len(lines)-1)*(zero,) + (one,)
+                logger.info('M_%s: f_{%s*m+%s} = %s * X_m',
+                            s_L, k**t_L, r_L, solution)
+                mu[s_L].append(solution)
+
+        d = len(seq(0)) + len(lines)
+        mu = tuple(Matrix(domain, [pad_right(tuple(row), d, zero=zero) for row in M])
+                         for M in mu)
+        right = vector(values(0, lines))
```

Comment:
I find it puzzling that the `values` function contains `seq` which in turn has the left vector of the given sequence times the matrix product, and this at the very end ends up in `right`.

https://github.com/sagemath/sage/issues/21204#issuecomment-1417820841 Item 16

Reply:
Acknowledged.

Reply:
Let me expand on that: the `right` vector of the bootstrapping sequence is never used. That is not a big issue, but is still interesting to observe (perhaps in an example?).

```
sage: R = kRegularSequenceSpace(2, ZZ)
sage: two = R.one_hadamard()*2
sage: two.linear_representation()
((1), Finite family {0: [1], 1: [1]}, (2))
sage: two_again = R.guess(lambda n: 2, sequence=two)
sage: two_again.linear_representation()
((2), Finite family {0: [1], 1: [1]}, (1))
```

And perhaps a comment here:

```suggestion
        right = vector(values(0, lines))
        # Note that the right vector of the given `sequence` is never
        # used. Instead, the corresponding left vector valued sequence
        # is used throughout when finding linear combinations, so the
        # same left vector valued sequence has to been used here for
        # consistency.
```

Reply:
This was indeed an issue (see other conversation). It changed now in c198d90dd02. Added the proposed example in 6aab949e4b1.

Reply:
ok, thank you.

Please resolve.